2.0-proposal/RELATIONS.md

```diff
@@ -0,0 +1,248 @@
+would be from `reactive/relations/myapp/peers.py`, in their respective layers.
+
+Also note that those instances would be available even if their relation is not
+related to any remote application.
+
+# The SimpleRelation API
+
+The `SimpleRelation` base class will provide a couple of automatically managed
+reactive flags, as well as providing collections for introspecting the state of
+the relation, including related applications and units, relation data provided
+by those units, and for associating reactive flags with particular related
+applications or units and filtering related applications and units by those
+flags.
+
+## Managed Flags
+
+The automatically managed flags are:
+
+  * `relations.{relation_name}.joined`
+  * `relations.{relation_name}.changed`
+
+where the `{relation_name}` is replaced at runtime with the appropriate
+relation name for that instance.  (The relation name for an instance can also
+be accessed as `relation_instance.relation_name`.)
+
+The `joined` flag will set as long as at least one remote unit is related via
+the relation, and will be updated whenever the list of related units changes so
+that any handler watching the flag will be re-invoked.
+
+The `changed` flag will be set as long as at least one remote unit is
+related and has set data on the relation, and it will be updated whenever
+either a unit updates its relation data or if a related unit goes away.  The
+interface can select only units whose relation data has changed using the
+`with_flag` filter method described below and this flag.
+
+## Application and Unit Collections
+
+The interface layer can iterate over related applications and units using the
+provided collections.  For example:
+
+```python
+class MyRelation(SimpleRelation):
+    class flags:
+        joined = 'relation.{relation_name}.joined'
+
+    @when(flags.joined)
+    def changed(self):
+        for app in self.applications:
+            for unit in app.units:
+                print('Unit {} has data: {}'.format(unit.name, unit.data))
+```
+
+Applications and units can have flags associated with them.  Flags that are
+added to an application are automatically added to all of that application's
+units, even ones that join after the fact, and if a flag is added to at least
+one of an application's units, that application will be considered to have that
+flag.
+
+You can filter the list of applications and units by flags.  For example:
+
+```python
+class MyRelation(SimpleRelation):
+    class flags:
+        changed = 'relation.{relation_name}.changed'
+        ready = 'relation.{relation_name}.ready'
+
+    @when(flags.changed)
+    def changed(self):
+        for app in self.applications:
+            for unit in app.units.with_flag(self.flags.changed):
+                host = unit.data.host
+                print('unit {} host changed: {}'.format(unit.name, host))
+                if host:
+                    unit.add_flag(self.flags.ready)
```

Comment:
[s/flag/state/] tl;dr: It's pretty clear that you are associating units/conversations with a flag. It's just not clear what "associating" does exactly. There is no intuitive explanation for the behavior of "associating" using the "flags" concept that users learn when using "normal" flags.

At least that's what tripped us up about conversations. `conversation.set_flag` does something completely different from `reactive.set_flag` and there is no conceptual explanation as to what the behavior is exactly. I had to try out a bunch of stuff and read through the documentation a couple of times to figure out the exact behavior.

  - `reactive.set_flag` adds a flag to the global scope that every layer can use in its handlers.
  - `reactive.remove_flag` removes that flag from the global scope.
  - `conv.set_flag` adds a special flag to the global scope. When handlers react to that flag *they get an object* containing all the conversations that have "added" this flag.
  - `conv.remove_flag` removes this conversation from *that object*. If no conversations are left, then the flag itself is also removed.

Even if you know how normal flags work, it's very hard to figure out how relationship flags work just by reading example code.

That's why I started talking about it as "you add a conversation to a flag" instead of "you add a flag to a conversation", because then you do have an overarching concept that explains both normal and relationship flags:

 - Flags contain objects. `reactive.set_flag` creates a flag with a "None" object.
 - Rename `conv.set_flag` to `conv.add_to_flag` and the function of this method is clear: it adds the conversation to the flag. Then it's also clear why `conv.remove_from_flag` doesn't remove the flag if there are conversations left in the flag.

Is that clear? That's why I was going for the "flags can contain object" thing, because that's for me a great way to explain the behavior of "associating a unit with a flag".